internal/librarian/createreleasepr.go

```diff
@@ -307,18 +307,35 @@ func generateReleaseCommitForEachLibrary(state *commandState, inputDirectory str
 	return pr, nil
 }
 
+// TODO(https://github.com/googleapis/librarian/issues/564): decide on release notes ordering
 func formatReleaseNotes(commitMessages []*CommitMessage) string {
-	features := []string{}
-	docs := []string{}
-	fixes := []string{}
+	// Group release notes by type, preserving ordering (FIFO)
+	var features, docs, fixes []string
+	featuresSeen := make(map[string]bool)
+	docsSeen := make(map[string]bool)
+	fixesSeen := make(map[string]bool)
 
-	// TODO(https://github.com/googleapis/librarian/issues/549): deduping (same message across multiple commits)
 	// TODO(https://github.com/googleapis/librarian/issues/547): perhaps record breaking changes in a separate section
 	// TODO(https://github.com/googleapis/librarian/issues/550): include backlinks, googleapis commits etc
 	for _, commitMessage := range commitMessages {
-		features = append(features, commitMessage.Features...)
-		docs = append(docs, commitMessage.Docs...)
-		fixes = append(fixes, commitMessage.Fixes...)
+		for _, feature := range commitMessage.Features {
```

Comment:
These are all the same, so this could be written
```
dedup := func(s []string) []string {
    seen := make(map[string]bool)
    var r []string
    for _, el := range s {
      if !seen[el] {.....}
   }
   return r
}

features := dedup(commitMessage.Features)
...
...
```

Reply:
Filed https://github.com/googleapis/librarian/issues/600 to consider following up on this